### PR TITLE
add the error message in the details box

### LINKED ISSUE #197: Add textarea to update origins for website
We need a textarea to allow people to enter their allowed origins urls.

### PR DESCRIPTION
**Pull request by Panioglo**
Fixes #197, #202, #204

### CODE REVIEW

File: app/controllers/websites_controller.rb [hunk 1]
```diff
@@ -6,7 +6,9 @@ def index
   end
 
   def update
-    @website.update_attributes(website_params)
+    strong_params = website_params
+    strong_params[:origins] = current_website.ensure_valid_protocol_for_origins(website_params[:origins])
+    @website.update_attributes(strong_params)
```
Comment (Chocksy) [inline app/controllers/websites_controller.rb]: Here we need to use some if clause to make sure that we check for validations. That way we can render the :edit action again if there is an error and show that error using simple_forms in the form.

Also that ensure_valid_protocol_for_origins should be as a callback in website to make sure that we always call it not here in the controller.

Use `before_validation :check_origins, if: :origins_changed?` something like this.

File: app/models/issue.rb [hunk 2]
```diff
@@ -12,6 +12,10 @@ def error
     ErrorStore.find(self)
   end
 
+  def data=(data)
+    super(ErrorStore::Utils.encode_and_compress(data))
+  end
+
```
Comment (Chocksy) [inline app/models/issue.rb:22]: here you should overwrite the `def data` method too to decode the column info. So it's always decoded when we access it anywhere in the codebase. That way you don't have to decode it every time you wanna read it!

File: app/models/website.rb [hunk 3]
```diff
@@ -66,7 +82,7 @@ def valid_origin?(value)
     return false if origins.blank? # no origins to check against
     return false if value.blank?
     value = value.downcase
-    origins.include?(value)
+    return !origins.include?(value)
```
Comment (Chocksy) [inline app/models/website.rb]: there is no need for return here. Use the rubocop styleguide please!. You add return only in the lines above but this is the last. The last line never needs a return it's understandable.

Comment (Chocksy) [inline app/models/website.rb]: ok after our discussion please revert this.

File: app/models/website.rb [hunk 4]
```diff
@@ -26,6 +26,22 @@ def website_dependent
     website_members.each(&:delete)
   end
 
+  def origins=(origins, string_origins: "", protocols: ["http", "https", "ftp", "ftps", "sftp"])
```
Comment (Chocksy) [inline app/models/website.rb]: don't do this it's not needed. You need a `before_validation :check_origins`. And in that check_origins method add errors with `errors.add(:origins, 'Please enter a valid origin')` that way when we update and save the website with new origins this error will show up.

File: app/models/website.rb [hunk 5]
```diff
@@ -18,14 +18,30 @@ class Website < ActiveRecord::Base
...
   before_create :generate_api_keys
   before_update :generate_api_keys, if: -> { generate }
+  before_update :check_required
   before_destroy :website_dependent
 
   def website_dependent
     website_members.each(&:delete)
   end
 
+  def check_origins( schemes: ["http://", "https://", "ftp://", "ftps://", "sftp://"] )
```
Comment (Chocksy) [inline app/models/website.rb]: stop setting default arguments for a method that will never be called elsewhere. Use a variable to define this default value inside the method.

Comment (Panioglo) [inline app/models/website.rb]: ok

File: spec/lib/error_store/manager_spec.rb [hunk 6]
```diff
@@ -113,9 +113,7 @@
       req = post_error_request(web_response_factory('ruby_exception'), website)
       data = validated_request(req)
       ErrorStore::Manager.new(data).store_error
-      expect {
-        subject
-      }.to change(Issue, :count).by(1)
+      expect(subject.group.present?).to eq(true)
```
Comment (Chocksy) [inline spec/lib/error_store/manager_spec.rb]: Why did you do this?

File: app/models/website.rb [hunk 7]
```diff
@@ -18,14 +18,31 @@ class Website < ActiveRecord::Base
...
   def website_dependent
     website_members.each(&:delete)
   end
 
+  def check_origins
+    return true if origins == '*'
+    schemes = ["http://", "https://", "ftp://", "ftps://", "sftp://"]
+    origins.split("\n").each do |origin|
+      unless schemes.any? { |scheme| origin.include? scheme }
+        errors.add(:origins, 'Please enter a valid origin')
```
Comment (Chocksy) [inline app/models/website.rb]: We need to be more specific with the error. For example in this case we are checking if the urls include a scheme. So the error should be `Please add origin urls with http/https etc.`

File: app/views/settings/details/_settings.html.haml [hunk 8]
```diff
@@ -9,19 +9,15 @@
           = f.input :domain, placeholder: 'http://www.domain.com'
           .dropdown
             = f.select :platform, @options, {}, prompt: '#{@website.platform}', :class => "btn btn-default cust-dropdown dropdown-toggle"
+        .form-group
+          = f.input :origins, as: :text, label: 'Allowed Domains', hint: "e.g. example.com or https://example.com", input_html: {:rows => "7"}, class: 'form-control'
```
Comment (Chocksy) [inline app/views/settings/details/_settings.html.haml]: Update the hint here as we are not accepting example.com without scheme.

File: spec/models/issue_spec.rb [hunk 9]
```diff
@@ -30,6 +30,8 @@
     :url=>"http://localhost///",
     :fragment=>nil}}
 
+  it { should be_kind_of(ErrorStore::Utils) }
```
Comment (Chocksy) [inline spec/models/issue_spec.rb]: should is not good. And what are you testing here that the issue is an Utils class? I would say to remove this line.

File: app/controllers/websites_controller.rb [hunk 10]
```diff
@@ -7,7 +7,11 @@ def index
 
   def update
     @website.update_attributes(website_params)
-    redirect_to settings_url(details_tab: 'settings', main_tab: 'details'), notice: 'Website updated'
+    unless @website.errors.full_messages.blank?
+      redirect_to settings_url, notice: @website.errors.full_messages.join(', ')
+    else
+      redirect_to settings_url, notice: 'Website updated'
+    end
```
Comment (Chocksy) [inline app/controllers/websites_controller.rb:13]: Change this to a more sane code:

```
if @website.update_attributes(website_params)
  redirect_to settings_url, notice: 'Website updated'
else
  redirect_to settings_url, notice: @website.errors.full_messages.join(', ')
end
```

Comment (Panioglo) [inline app/controllers/websites_controller.rb:13]: :)

**Pull request merged**